Mikem 1081 telemetry.

mmastand opened the pull request:
This required what looks like a large refactor to work in a tidy way (no sinks, files to read/write, or `on.exit`s). That all fell apart on travis/app and I think they just pointed out how fragile it was. Plus, it wasn't possible to return a value on error. If write log is TRUE, predict will now return telemetry data with success or failure and write the log.

Big Changes:
- Added model name param to `machine learn`, `tune`, and `flash`. Defaults to the name of the outcome column.
- All code from `predict.model_list` was moved to `predict_model_list_main`. Predict.model_list handles 4 use cases by dispatching to the correct function.
    - `log = FALSE`: runs code exactly as it used to be (but is not contained in `predict_model_list_main`) and returns prediction df with telemetry as an attribute df. If there is an error, it will error out as normal.
    - `log = TRUE`: `predict_model_list_main returns` is wrapped in a `safe_n_quiet` call to capture any error messages or warnings. After execution, `parse_safe_n_quiet` will arrange outputs properly. A user will always get an updated log file and predictions df with telemetry tibble as an attribute. If the function errors, a loud warning will be output to the console, the predictions dataframe will have 0 rows, and error info will be in the telemetry table.

Minor Changes:
- Added more info to the log file.
- Log file returns telemetry dataframe as an attribute to `newdata`
- If loaded from `load_models`, RDS filename will be saved. Otherwise, `loaded_from` is set to `trained_in_memory`.
- `print.model_list` method prints model name.

Added tests for model name, logging, and print.

mmastand commented:
I'm pushing more error handling code into this. Don't feel like you need to review just yet.

michaellevy commented:
@mmastand make sure this fits well with @taylorlarsen 's vision around [shared schemas](https://github.com/HealthCatalyst/ml.internal/issues/1092)

mmastand commented:
@michaellevy Thanks for the review. A couple comments and discussion points. I put time on your calendar for Westchester, maybe we can talk through this as well.
*If a model_name is provided to the training function, but a filepath isn't provided to save_models, let's use {model_name}.RDS instead of models.RDS as the filename.*
	The model will always have a model name. The `model_name` is either user-specified or the name of the outcome column. I think we should leave it as models.RDS if unspecified.

*Warnings from predict should be captured and written to the log file.*
	`safely` will return [errors or results]. `quietly` will return warnings, messages, output, and results. Not errors though. Do you think the best way to do this is `safely(quietly(predict))`. That feels like a lot of logic after predict to me. Can you think of a better way?

Jog my memory on how to get the output data's structure from the training data? Without unprepping, I'm not seeing it. Data gets transformed all the way through predict.
```
library(healthcareai)
#> healthcareai version 2.0.0
#> Please visit https://docs.healthcare.ai for full documentation and vignettes. Join the community at https://healthcare-ai.slack.com
m <- machine_learn(pima_diabetes[1:50, 6:10], outcome = diabetes, models = "rf",
model_name = "telemetry_test")
#> Training new data prep recipe
#> diabetes looks categorical, so training classification algorithms.
#> Training with cross validation: Random Forest
#> Warning: package 'caret' was built under R version 3.4.3
#> 
#> *** Models successfully trained. The model object contains the training data minus ignored ID columns. ***
#> *** If there was PHI in training data, normal PHI protocols apply to the model object. ***
p <- predict(object = m, newdata = pima_diabetes[1:50, 6:10],
write_log = TRUE, prepdata = TRUE)
#> Prepping data based on provided recipe

p_error <- m$`Random Forest`$trainingData %>%
dplyr::mutate(diabetes = .outcome) %>% # Simulates what happens in predict.
dplyr::select(- .outcome)
head(p_error)
#>   insulin pedigree age weight_class_normal weight_class_obese
#> 1  203.85    0.627  50                   0                  1
#> 2  203.85    0.351  31                   0                  0
#> 3  203.85    0.672  32                   1                  0
#> 4   94.00    0.167  21                   0                  0
#> 5  168.00    2.288  33                   0                  0
#> 6  203.85    0.201  30                   0                  0
#>   weight_class_overweight weight_class_missing weight_class_other diabetes
#> 1                       0                    0                  0        Y
#> 2                       1                    0                  0        N
#> 3                       0                    0                  0        Y
#> 4                       1                    0                  0        N
#> 5                       0                    0                  0        Y
#> 6                       1                    0                  0        N
head(p)
#> # A tibble: 6 x 6
#>   diabetes predicted_diabetes insulin weight_class   pedigree   age
#>   <chr>                 <dbl>   <int> <chr>             <dbl> <int>
#> 1 Y                     0.610      NA obese             0.627    50
#> 2 N                     0.399      NA overweight        0.351    31
#> 3 Y                     0.380      NA normal            0.672    32
#> 4 N                     0.358      94 overweight        0.167    21
#> 5 Y                     0.600     168 morbidly obese    2.29     33
#> 6 N                     0.395      NA overweight        0.201    30
```

taylorlarsen commented:
@mmastand I played with `predict()` with `write_log = TRUE` a bit. Super helpful functionality, thanks!

My first pass at comments/observations:

- It might be good to add `attr(test_predictions,"prediction_log")` to the documentation so that it's more obvious to the user how to extract the log from the predictions.

- Looking at the Shared.Model.ModelTelemetry, the only thing that is missing from "prediction_log" and can't be derived from the existing information is `LastRunTimeNBR` which may not be intuitively named but I was planning for it to represent the amount of time in seconds that predict spent executing.

- It would be convenient if `n_predictions` from "prediction_log" was set to zero if no predictions were made instead of NA as we'd otherwise likely end up doing that transformation later in the process anyways.

- If `predict` fails because the model doesn't exist, a log is never written. The use-case I'm picturing here is if the model is somehow moved/deleted/renamed after deployment, we'd want to know about it via telemetry.

michaellevy commented:
Warnings and errors when write_log is TRUE look good, both in the console and the log file. Thanks for that. There are some git problems that need to be fixed (at the bottom of this comment). I like @taylorlarsen's comments; however, I don't think there's anything we can do if the model object doesn't exist since `predict.model_list` isn't dispatched then. Maybe you just have to know that if the log isn't written it means the model wasn't found. 

A couple other minor comments, then this is good to go by me.

- Update the documentation, `write_log` and `return` especially, to be clear about when you might get a zero-row DF back, when errors won't be errors, etc. Also add a description of the `failed` attribute.
- Rounding in `update_telemetry` rounds the underlying values. This should only happen at print time, if ever. This bit me recently... missingness used to round, and I couldn't figure out why I was getting missing value errors in training when missingness returned all zeros, but in fact a fraction of a percent of rows were missing values.
- `stop_prod_logs` and `catalyst_test_deploy_in_prod` should be marked as defunct rather than deprecated since they no longer work. You could use the `.Defunct` function. Helpful: https://stackoverflow.com/a/10145627/2565816
- The `model_name` arg to flash_models and others should default to either nothing or NULL instead of NA? I'm guessing you did this because passing it through to `as.model_list` as a blank argument doesn't work. I'd rather NULL than NA.
	- Update `machine_learn` (and possibly elsewhere) docs that say "if left blank..."

Git stuff:
- This appears to revert changes I made to README in the most recent code I merged. 
- There's a merge conflict still present in docs/pkgdown.yml, and there are changes to files in docs/. 
- It looks like some search files got deleted: docs/docsearch.css, docs/docsearch.json, docs/sitemap.xml
- There are some .RDS files in the root directory that shouldn't be under version control

mmastand commented:
Thanks for the reviews! I'll merge once checks all pass.

Larsens:

- Added instructions on how to extract prediction_log
- Added `run_time`, in seconds.
- Set to 0
- Nothing we can do...

Levy:

- Updated predict docs in the write_log, return, and details
- Removed rounding in update_telemetry
- Changed to `.Defunct`
- Correct! Using `NULL` now.
- Updated docs to say, "Defaults to the name of the outcome variable."

Git:

- Replaced readme with master. 0 changed lines.
- Replaced docs folder with master. 0 changed lines in every file.
- Removed extra RDS files.
